Add curl_getinfo dynamic return type extension

PrinsFrank opened the pull request.

PrinsFrank committed:
Add Dynamic return type extension for curl_getinfo

PrinsFrank committed:
register new dynamic return type extension

PrinsFrank committed:
Remove documentation for _T constants

PrinsFrank committed:
Add test cases

staabm commented on src/Type/Php/CurlGetinfoFunctionDynamicReturnTypeExtension.php:
```diff
@@ -0,0 +1,247 @@
+
+	public function isFunctionSupported(FunctionReflection $functionReflection): bool
+	{
+		return $functionReflection->getName() === 'curl_getinfo';
```
this should be case-insensitivd… usually we use strtolower() here

PrinsFrank committed:
Fix code style

staabm commented on src/Type/Php/CurlGetinfoFunctionDynamicReturnTypeExtension.php:
```diff
@@ -0,0 +1,247 @@
+
+	private function cacheReturnTypes(): void
+	{
+		if (isset($this->componentTypesPairedConstants) === true) {
```
style-wise, I would move this check to the call-site of this method

staabm commented on src/Type/Php/CurlGetinfoFunctionDynamicReturnTypeExtension.php:
```diff
@@ -0,0 +1,247 @@
+		$arrayOfIntegerKeyedStrings = new ArrayType($integer, $string);
+		$arrayOfArrayIntegerKeyedStrings = new ArrayType($integer, $arrayOfIntegerKeyedStrings);
+
+		$this->componentTypesPairedConstants = [
```
Is this caching even necessary..? Do you have evidence it actually prevent a bottleneck?

PrinsFrank replied:
I might have used another class as inspiration, so I thought it was required. I removed the caching in a new commit.

PrinsFrank committed:
Accept any casing of the curl_getinfo function

staabm commented on src/Type/Php/CurlGetinfoFunctionDynamicReturnTypeExtension.php:
```diff
@@ -0,0 +1,247 @@
+use PHPStan\Type\Type;
+use PHPStan\Type\TypeCombinator;
+use function count;
+use const CURLINFO_APPCONNECT_TIME;
```
Instead of adding such `use`statements, which render phpstan dependent on the curl extension, you should use the `ReflectionProvider` instead and use `hasConstant`/`getConstant`

PrinsFrank committed:
Remove type caching

PrinsFrank committed:
Don't rely on constants from ext-curl but use reflectionProvider instead

PrinsFrank committed:
Fix phpstan-static-reflection failures

PrinsFrank committed:
Remove constants that don't work in older PHP versions

PrinsFrank committed:
Remove tests for constants that don't work in older PHP versions

staabm commented:
When you use https://github.com/phpstan/phpstan-src/blob/1.5.x/tests/PHPStan/Analyser/NodeScopeResolverTest.php for testing instead of the Legacy test-class, you can add separate test files per php-version, so you don‘t need to limit this feature on constants available on all versions

PrinsFrank commented:
@staabm Thanks for the fast review and all the help so far! I will finalize the tests tomorrow!

PrinsFrank committed:
Add remaining constants introduced with PHP7.3

staabm commented on src/Type/Php/CurlGetinfoFunctionDynamicReturnTypeExtension.php:
```diff
@@ -0,0 +1,202 @@
+				throw new ShouldNotHappenException();
+			}
+
+			if ($valueType->getValue() === $componentType->getValue()) {
```
I guess this should be

```suggestion
			if ($componentType->getValue()->isSuperTypeOf($valueType)->yes()) {
```

or maybe the other way arround

staabm commented on src/Type/Php/CurlGetinfoFunctionDynamicReturnTypeExtension.php:
```diff
@@ -0,0 +1,202 @@
+			}
+
+			$valueType = $this->reflectionProvider->getConstant($constantNameNode, $scope)->getValueType();
+			if ($valueType instanceof ConstantIntegerType === false) {
```
are all these constants integers? if not, you might use `ConstantScalarType ` instead

staabm commented on src/Type/Php/CurlGetinfoFunctionDynamicReturnTypeExtension.php:
```diff
@@ -0,0 +1,202 @@
+		if (count($functionCall->getArgs()) > 1) {
+			$componentType = $scope->getType($functionCall->getArgs()[1]->value);
+
+			if (!$componentType instanceof ConstantType) {
```
not sure what this IF case is trying to detect

staabm reviewed (commented):
great job so far.

for a few days there is also a 1.4.x branch.. I think this PR should target that branch instead of the longer termin 1.5.x

PrinsFrank committed:
Move tests from LegacyNodeScopeResolverTest to NodeScopeResolverTest and add php version specific tests

PrinsFrank committed:
Use isSuperTypeOf and remove subsequent unnecessary shouldNotHappenException and process further feedback

PrinsFrank committed:
Move test yields to some lines above because csfixer wants to return early here but that makes the test less maintainable

PrinsFrank committed:
Readd check for ConstantIntegerType as getValue method is not present on base Type class

PrinsFrank commented:
@staabm Thanks! I rebased on 1.4.x, moved the tests to the non-legacy test cases and processed the rest of the feedback. The remaining 2 checks are unrelated to the changes in this branch, right? Should I try to fix them in this branch?

PrinsFrank committed:
Move scenarios into method to prevent PHPUnit running them

staabm commented:
From my perspective this looks good, thank you.

we need to double check the errors in https://github.com/phpstan/phpstan-src/runs/5559798589?check_suite_focus=true

otherwise - lets wait for some feedback from ondrey

staabm commented on tests/PHPStan/Analyser/data/curl_getinfo_7.3.php:
```diff
@@ -0,0 +1,31 @@
+	{
+		$handle = new CurlHandle();
+		assertType('float', curl_getinfo($handle, CURLINFO_CONTENT_LENGTH_DOWNLOAD_T));
+		assertType('float', curl_getinfo($handle, CURLINFO_CONTENT_LENGTH_UPLOAD_T));
```
Are these 2 length fields integers?

staabm commented on tests/PHPStan/Analyser/data/curl_getinfo_7.3.php:
```diff
@@ -0,0 +1,31 @@
+		assertType('float', curl_getinfo($handle, CURLINFO_SPEED_UPLOAD_T));
+		assertType('float', curl_getinfo($handle, CURLINFO_APPCONNECT_TIME_T));
+		assertType('float', curl_getinfo($handle, CURLINFO_CONNECT_TIME_T));
+		assertType('float', curl_getinfo($handle, CURLINFO_FILETIME_T));
```
Timestamp… int?

PrinsFrank committed:
Speed and sizes are integers

PrinsFrank committed:
Speed and sizes are integers

PrinsFrank committed:
Speed and sizes are integers

ondrejmirtes commented on src/Type/Php/CurlGetinfoFunctionDynamicReturnTypeExtension.php:
```diff
@@ -0,0 +1,198 @@
+
+	public function isFunctionSupported(FunctionReflection $functionReflection): bool
+	{
+		return strtolower($functionReflection->getName()) === 'curl_getinfo';
```
No need to strtolower here

ondrejmirtes commented on src/Type/Php/CurlGetinfoFunctionDynamicReturnTypeExtension.php:
```diff
@@ -0,0 +1,198 @@
+
+			$componentType = $componentType->toInteger();
+			if (!$componentType instanceof ConstantIntegerType) {
+				throw new ShouldNotHappenException();
```
Wrong calls with a "string" or similar should not throw an exception - that leads to internal error. Please test this in `tests/PHPStan/Analyser/data/curl_getinfo.php` and return the default return type instead.

ondrejmirtes commented on src/Type/Php/CurlGetinfoFunctionDynamicReturnTypeExtension.php:
```diff
@@ -0,0 +1,198 @@
+			'CURLINFO_CONNECT_TIME' => new FloatType(),
+			'CURLINFO_PRETRANSFER_TIME' => new FloatType(),
+			'CURLINFO_STARTTRANSFER_TIME' => new FloatType(),
+			'CURLINFO_REDIRECT_COUNT' => new IntegerType(),
```
Please be more memory usage friendly, assign variables with these types and then use the variables instead of creating new instance every time.

PrinsFrank committed:
Remove superfluous strtolower

PrinsFrank committed:
Return allComponentsReturnType when the second parameter of curl_getinfo is null

PrinsFrank committed:
Remove unused function import

PrinsFrank committed:
Add test cases for invalid second arguments for curl_getinfo and return the correct type

PrinsFrank committed:
Preserve memory by not constructing unnecessary duplicate objects

PrinsFrank committed:
Return correct type for non constant values

ondrejmirtes commented:
Thank you! I merged it in 1.5.x: https://github.com/phpstan/phpstan-src/commit/ef59f08050742186ef38d816ea6153977b5ba571

ondrejmirtes commented:
FYI I checked PHP docs and concluded that the keys are not optional: https://github.com/phpstan/phpstan-src/commit/9cde41375bc3dde9516ce3af2fed0143f3bfbf94
